discovery_engine_core/core/src/engine.rs

```diff
@@ -324,35 +324,60 @@ where
         };
 
         let mut errors = Vec::new();
-        for stack in &mut stacks {
-            let articles = stack
-                .new_items(&key_phrases)
-                .await
-                .and_then(|articles| stack.filter_articles(history, articles));
-
-            match articles.map_err(Error::StackOpFailed).and_then(|articles| {
-                let id = stack.id();
-                articles
-                    .into_iter()
-                    .map(|article| {
-                        let title = article.title.as_str();
-                        let embedding = self.ranker.compute_smbert(title).map_err(Error::Ranker)?;
-                        document_from_article(article, id, embedding).map_err(Error::Document)
-                    })
-                    .collect::<Result<Vec<_>, _>>()
-            }) {
-                Ok(documents) => {
-                    if let Err(error) = stack.update(&documents, &mut self.ranker) {
-                        errors.push(Error::StackOpFailed(error));
-                    } else {
-                        stack.data.retain_top(self.core_config.keep_top);
-                    }
+        let max_errors = stacks.len();
+        for stack in stacks {
+            let articles = match stack.new_items(&key_phrases).await {
+                Ok(articles) => articles,
+                Err(error) => {
+                    let error = Error::StackOpFailed(error);
+                    error!("{}", error);
+                    errors.push(error);
+                    continue;
+                }
+            };
+
+            let articles = match stack.filter_articles(history, articles) {
+                Ok(articles) => articles,
+                Err(error) => {
+                    let error = Error::StackOpFailed(error);
+                    error!("{}", error);
+                    errors.push(error);
+                    continue;
                 }
-                Err(error) => errors.push(error),
+            };
+
+            let id = stack.id();
+            let (documents, article_errors) = articles
+                .into_iter()
+                .map(|article| {
+                    let title = article.title.as_str();
+                    let embedding = self.ranker.compute_smbert(title).map_err(|error| {
+                        let error = Error::Ranker(error);
+                        error!("{}", error);
+                        error
+                    })?;
+                    document_from_article(article, id, embedding).map_err(|error| {
+                        let error = Error::Document(error);
+                        error!("{}", error);
+                        error
+                    })
+                })
+                .partition_result::<Vec<_>, Vec<_>, _, _>();
+            if documents.is_empty() {
```

Comment:
This is not necessary an error, if `new_items` returns no articles or if the filter remove all of them we will have no documents and no errors.